### PR TITLE
stake-pool: CLI QOL improvements

### PR DESCRIPTION
**Pull request by gabrielhicks**
When creating a stake-pool using the CLI, if at any point an instruction fails subsequent commands will also fail. This behavior creates a problem when initializing a stake pool in the CLI with vanity keypairs. This solution aims to break up the `command_create_pool` into iterative steps with checks along the way.

I had changed the deprecated ` use spl_associated_token_account::create_associated_token_account` but am happy to change it back.

### CODE REVIEW

File: stake-pool/cli/src/main.rs [hunk 1]
```diff
@@ -444,50 +567,121 @@ fn command_create_pool(
...
+        &mint_keypair,
+        &pool_fee_account,
+        deposit_authority,
+        epoch_fee,
+        withdrawal_fee,
+        deposit_fee,
+        referral_fee,
+        max_validators,
+        &withdraw_authority,
+    )?;
```
Comment (joncinque) [inline stake-pool/cli/src/main.rs:683]: Can we combine `setup_validator_list_account` and `initialize_stake_pool` into the same transaction? Fewer transactions is always best. And although the possibility is really remote, it's technically possible for someone to steal a created-but-not-initialized account, since no data has been written to it yet

File: stake-pool/cli/src/main.rs [hunk 2]
```diff
@@ -374,66 +344,219 @@ fn command_create_pool(
...
+    if let Ok(account) = stake_pool_account_info {
+        let stake_pool_manager = get_stake_pool(&config.rpc_client, &stake_pool_keypair.pubkey()).unwrap().manager;
+        if account.owner == spl_stake_pool::id() {
+            if stake_pool_manager == config.manager.pubkey() {
+                println!("Stake pool account {} already exists and is initialized", stake_pool_keypair.pubkey());
+                return Ok(());
+            } else {
+                println!("Stake pool account {} already exists but is not initialized", stake_pool_keypair.pubkey());
+            }
+        }
```
Comment (joncinque) [inline stake-pool/cli/src/main.rs:520]: This could unnecessarily panic -- rather than checking the manager, we can just check that the first byte in the data is equal to `AccountType::StakePool as u8` (if there is any data in the account) https://github.com/solana-labs/solana-program-library/blob/c4d51f20b8722d23b03e94c7075702fca52460dd/stake-pool/program/src/state.rs#L37

File: stake-pool/cli/src/main.rs [hunk 3]
```diff
@@ -374,66 +344,219 @@ fn command_create_pool(
...
+        ];
+
+        unique_signers!(create_signers);
+        let create_transaction = checked_transaction_with_signers(config, &create_instructions, &create_signers)?;
+
+        println!(
+            "Creating stake pool account {}",
+            stake_pool_keypair.pubkey(),
+        );
+        send_transaction(config, create_transaction)?;
```
Comment (joncinque) [inline stake-pool/cli/src/main.rs:545]: Let's put the creation and initialization of the stake pool in the same transaction (along with the creation of the validator list, as mentioned in another comment) -- more transactions just means more opportunities for things to go wrong

File: stake-pool/cli/src/main.rs [hunk 4]
```diff
@@ -2,9 +2,9 @@
 mod client;
 mod output;
 
-// use instruction::create_associated_token_account once ATA 1.0.5 is released
-#[allow(deprecated)]
-use spl_associated_token_account::create_associated_token_account;
+use spl_associated_token_account::instruction::create_associated_token_account; // once ATA 1.0.5 is released
```
Comment (joncinque) [inline stake-pool/cli/src/main.rs:5]: This is great, thanks! Normally I'd ask for it to go in a separate PR, but it's a small enough change. You could even remove the comment
```suggestion
use spl_associated_token_account::instruction::create_associated_token_account;
```

File: stake-pool/cli/src/main.rs [hunk 5]
```diff
@@ -2,9 +2,9 @@
...
+// #[allow(deprecated)]
+// use spl_associated_token_account::create_associated_token_account;
```
Comment (joncinque) [inline stake-pool/cli/src/main.rs:7]: nit: Feel free to remove these comments

File: stake-pool/cli/src/main.rs [hunk 6]
```diff
@@ -775,11 +969,12 @@ fn add_associated_token_account(
             .get_minimum_balance_for_rent_exemption(spl_token::state::Account::LEN)
             .unwrap();
 
-        #[allow(deprecated)]
+        // #[allow(deprecated)]
```
Comment (joncinque) [inline stake-pool/cli/src/main.rs:972]: nit: Feel free to remove this comment too

**Review by joncinque — COMMENTED** [review]
Thanks for the contribution! This will definitely make the experience of creating a stake pool much smoother. Let me know if anything's unclear

**Comment by gabrielhicks** [comment]
> Thanks for the contribution! This will definitely make the experience of creating a stake pool much smoother. Let me know if anything's unclear

I believe I have walked through each change and implemented! I appreciate the feedback and am happy to continue iterating if needed.

File: stake-pool/cli/src/main.rs [hunk 7]
```diff
@@ -374,120 +340,304 @@ fn command_create_pool(
...
-        // Validator stake account list storage
-        system_instruction::create_account(
+    let signers = vec![config.fee_payer.as_ref()];
+    let transaction = checked_transaction_with_signers(config, &instructions, &signers)?;
+
+    send_transaction(config, transaction)?;
+    Ok(())
+}
+
+fn setup_and_initialize_validator_list_with_stake_pool(
```
Comment (joncinque) [inline stake-pool/cli/src/main.rs:447]: This needs a clippy allow according to CI
```suggestion

#[allow(clippy::too_many_arguments)]
fn setup_and_initialize_validator_list_with_stake_pool(
```

**Review by joncinque — COMMENTED** [review]
Thanks for addressing all of these! Just a few last comments before merging.

I've kicked off CI, and it's failing the formatting check -- can you run `./cargo-nightly fmt --all` from the top of the repo?

You'll also need to run `./cargo-nightly clippy -Zunstable-options --all-targets -- --deny=warnings --deny=clippy::arithmetic_side_effects` and make sure everything's ok there. I think it'll be OK after addressing the remaining comments.

File: stake-pool/cli/src/main.rs [hunk 8]
```diff
@@ -374,120 +340,304 @@ fn command_create_pool(
...
             &config.fee_payer.pubkey(),
             &stake_pool_keypair.pubkey(),
             stake_pool_account_lamports,
             get_packed_len::<StakePool>() as u64,
             &spl_stake_pool::id(),
-        ),
-        // Initialize stake pool
-        spl_stake_pool::instruction::initialize(
+        ));
+        instructions.push(spl_stake_pool::instruction::initialize(
```
[the remark this answers is not in the document]
Comment (gabrielhicks) [reply stake-pool/cli/src/main.rs:514]: I was wondering if I should even have checks here as my assumption and understanding is if you are making it this far, we want you to do all of the ixs together

[the remark this answers is not in the document]
Comment (joncinque) [reply stake-pool/cli/src/main.rs:514]: Yeah that's true, but who knows, maybe someone wants to initialize one of the accounts separately themselves? It can't hurt, thankfully

**Comment by gabrielhicks** [comment]
> Thanks for addressing all of these! Just a few last comments before merging.
> 
> I've kicked off CI, and it's failing the formatting check -- can you run `./cargo-nightly fmt --all` from the top of the repo?
> 
> You'll also need to run `./cargo-nightly clippy -Zunstable-options --all-targets -- --deny=warnings --deny=clippy::arithmetic_side_effects` and make sure everything's ok there. I think it'll be OK after addressing the remaining comments.

Ran with changes and all looks good! Thank you!

**Review by joncinque — APPROVED** [review]
Thanks again for your contribution!
